### PR TITLE
Attribute exclusion in product export

### PR DESCRIPTION
**Pull request by axel-ducret**
Added ability to select attributes and associations not to export in product export

### CODE REVIEW

File: EventSubscriber/JobProfileSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            JobProfileEvents::POST_EDIT => 'onPostEdit'
+        );
+    }
+
+    /**
+     * Executed after job edit (just before rendering) : redirects to the custom template
+     *
+     * @param GenericEvent $event
+     */
+    public function onPostEdit($event)
```
Comment (BitOne) [inline EventSubscriber/JobProfileSubscriber.php]: Can you typehint the $event, please ?

File: EventSubscriber/JobProfileSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $jobInstance = $event->getSubject();
+
+        /** @var JobInstance $jobInstance */
```
Comment (BitOne) [inline EventSubscriber/JobProfileSubscriber.php]: Can avoid this kind of IDE hint in your code, as it's IDE specific stuff, and not phpDoc.

File: Processor/ProductToFlatArrayProcessor.php [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        parent::__construct($serializer, $channelManager, $mediaAttributeTypes);
+
+        $this->attributeGroupRepository  = $attributeGroupRepository;
+        $this->associationTypeRepository = $associationTypeRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process($product)
```
Comment (BitOne) [inline Processor/ProductToFlatArrayProcessor.php:68]: I think the best strategy would be to simply call `parent::process()` and then filter out the produced array to remove the unwanted field.

This way, you keep the normal behavior of the processor, and benefit from all the new stuff that are coming with new versions.

File: Processor/ProductToFlatArrayProcessor.php [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+     *
+     * @return array
+     */
+    protected function getAttributeList()
+    {
+        $attributeGroups = $this->attributeGroupRepository->findAll();
+
+        $results = array();
+        foreach ($attributeGroups as $attributeGroup) {
+            /** @var AttributeGroup $attributeGroup */
```
Comment (BitOne) [inline Processor/ProductToFlatArrayProcessor.php]: Same comment than before

File: Processor/ProductToFlatArrayProcessor.php [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                foreach ($attributes as $attribute) {
+                    $results[$attributeGroup->getLabel()][$attribute->getCode()] = $attribute->getLabel();
+                }
+            }
+        }
+        return $results;
+    }
+
+    /**
+     * Gets the list of all existing associations
```
Comment (BitOne) [inline Processor/ProductToFlatArrayProcessor.php]: It's more about association type than associations.

File: Processor/ProductToFlatArrayProcessor.php [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                'type'      => 'choice',
+                'options'   => [
+                    'choices'   => $this->getAttributeList(),
+                    'select2'   => true,
+                    'multiple'  => true,
+                    'label'     => 'pim_enhanced_connector.product_processor.attributes_to_exclude.label',
+                    'help'      => 'pim_enhanced_connector.product_processor.attributes_to_exclude.help'
+                ]
+            ],
+            'associationsToExclude' => [
```
Comment (BitOne) [inline Processor/ProductToFlatArrayProcessor.php]: It's more about associationTypes than associations themselves.

File: Processor/ProductToFlatArrayProcessor.php [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        $filteredMediaValues = array();
+
+        foreach ($mediaValues as $mediaValue) {
+            if (!in_array($mediaValue->getAttribute()->getCode(), $this->attributesToExclude)) {
+                $filteredMediaValues[] = $mediaValue;
+            }
+        }
+        return $filteredMediaValues;
+    }
+}
```
Comment (BitOne) [inline Processor/ProductToFlatArrayProcessor.php]: Missing return at last line.

File: Resources/config/assets.yml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+css:
+    pim_enhanced_connector:
+        - bundles/pimenhancedconnector/css/select2.multiselect.groups.css
```
Comment (BitOne) [inline Resources/config/assets.yml]: Idem, missing return character at the end of last line.

**Comment by BitOne** [comment]
Hey @axel-ducret !

Thanks for your contribution ! Great work ! Some cosmetic remarks and one architectural one on the processor, but otherwise, should be ok once they are fixed.

**Comment by axel-ducret** [comment]
Hey @BitOne !

Here are the corrections

File: Processor/ProductToFlatArrayProcessor.php [hunk 9]
```diff
@@ -17,26 +22,44 @@
...
+     * @param AttributeGroupRepository     $attributeGroupRepository
+     * @param AssociationTypeRepository    $associationTypeRepository
      */
     public function __construct(
         Serializer $serializer,
         ChannelManager $channelManager,
         array $mediaAttributeTypes,
         ProductBuilderInterface $productBuilder = null,
-        ObjectDetacherInterface $objectDetacher = null
+        ObjectDetacherInterface $objectDetacher = null,
```
Comment (axel-ducret) [inline Processor/ProductToFlatArrayProcessor.php:54]: Why is the objectDetacher optionnal ? Should I make my parameters optionnal too ?

**Comment by fitn** [comment]
Hi @axel-ducret,

Sorry for this late answer but we did not forget you ^^

We can't merge your PR for now. We need to check the scalability of this new feature as this bundle is used by customers with big catalog. I think this code works well on hundred of attributes but I'm pretty sure that it could not work with thousands (as findAll method is called).

However, I want to thank you for your contribution because it is a really needed (and wanted) feature and you did a great job. We will do the best to merge it (probably by the end of the month) ;)

Regards,

**Comment by jmleroux** [comment]
Hi @axel-ducret ,

we just forked your PR to rebase it before we can validate it.
Thanks again for this great contribution.

JM

**Comment by jmleroux** [comment]
@axel-ducret , i've rebased your branch on master in #67 to be able to validate this great PR.
Thank you again.

**Pull request closed without merging**